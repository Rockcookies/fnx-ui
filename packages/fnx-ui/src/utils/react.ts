import {
	Children,
	forwardRef,
	ForwardRefExoticComponent,
	ForwardRefRenderFunction,
	FunctionComponent,
	PropsWithoutRef,
	ReactElement,
	RefAttributes,
} from 'react';
import { isFragment } from 'react-is';

export const toElementArray = (
	children: React.ReactNode,
	option?: { keepEmpty?: boolean },
): ReactElement[] => {
	const keepEmpty = option && option.keepEmpty;
	let ret: React.ReactElement[] = [];

	Children.forEach(children, (child: any) => {
		if ((child === undefined || child === null) && !keepEmpty) {
			return;
		}

		if (Array.isArray(child)) {
			ret = ret.concat(toElementArray(child));
		} else if (isFragment(child) && child.props) {
			ret = ret.concat(toElementArray(child.props.children, option));
		} else {
			ret.push(child);
		}
	});

	return ret;
};

export const createForwardRef = <R, P = any>(
	displayName: string,
	render: ForwardRefRenderFunction<R, P>,
): ForwardRefExoticComponent<PropsWithoutRef<P> & RefAttributes<R>> => {
	const fc = forwardRef(render);
	fc.displayName = displayName;
	return fc;
};

export const createDefaultsForwardRef = <R, P, D>(
	displayName: string,
	defaultProps: D,
	render: ForwardRefRenderFunction<R, P & D>,
): ForwardRefExoticComponent<
	PropsWithoutRef<P & Partial<D>> & RefAttributes<R>
> => {
	const fc = forwardRef(render);
	fc.displayName = displayName;
	fc.defaultProps = defaultProps;
	return fc as any;
};

export const createFC = <P = any>(
	displayName: string,
	fc: FunctionComponent<P>,
): FunctionComponent<P> => {
	fc.displayName = displayName;
	return fc;
};

export const createDefaultsFC = <P, D>(
	displayName: string,
	defaultProps: D,
	fc: FunctionComponent<P & D>,
): FunctionComponent<P & Partial<D>> => {
	fc.displayName = displayName;
	fc.defaultProps = defaultProps;
	return fc as any;
};
