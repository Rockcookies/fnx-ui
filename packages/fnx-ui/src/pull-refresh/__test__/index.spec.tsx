import { act, fireEvent, render, waitFor } from '@testing-library/react';
import React from 'react';
import TestsDOM from '../../../test/dom';
import TestsEvent from '../../../test/events';
import PullRefresh from '../index';

describe('<PullRefresh/>', () => {
	beforeEach(() => {
		jest.useFakeTimers();
	});

	afterEach(() => {
		jest.useRealTimers();
	});

	it('should render different head content in different pulling status', async () => {
		const onRefresh = jest.fn();

		const { container, rerender } = render(
			<PullRefresh onRefresh={onRefresh} />,
		);

		const track = TestsDOM.mustQuerySelector(
			container,
			'.fnx-pull-refresh__track',
		);

		// pulling
		await TestsEvent.triggerTouch(track, 'touchstart', [[0, 0]]);
		await TestsEvent.triggerTouch(track, 'touchmove', [[0, 20]]);

		expect(container).toMatchSnapshot();

		// loosing
		await TestsEvent.triggerTouch(track, 'touchmove', [[0, 100]]);

		expect(container).toMatchSnapshot();

		// loading
		await TestsEvent.triggerTouch(track, 'touchend', [[0, 100]]);

		expect(container).toMatchSnapshot();

		// still loading
		await TestsEvent.triggerDrag(track, [0, 100]);

		expect(container).toMatchSnapshot();

		expect(onRefresh).toHaveBeenCalled();

		// end loading
		act(() => {
			rerender(<PullRefresh refreshing={true} />);
		});

		act(() => {
			rerender(<PullRefresh refreshing={false} />);
		});

		expect(container).toMatchSnapshot();
	});

	it('should render status slots correctly', async () => {
		const { container } = render(
			<PullRefresh
				slots={{
					pulling: (distance) => `pulling ${distance}`,
					loosing: (distance) => `loosing ${distance}`,
					loading: (distance) => `loading ${distance}`,
				}}
			/>,
		);

		const track = TestsDOM.mustQuerySelector(
			container,
			'.fnx-pull-refresh__track',
		);

		// pulling
		await TestsEvent.triggerTouch(track, 'touchstart', [[0, 0]]);
		await TestsEvent.triggerTouch(track, 'touchmove', [[0, 20]]);

		expect(container).toMatchSnapshot();

		// loosing
		await TestsEvent.triggerTouch(track, 'touchmove', [[0, 100]]);

		expect(container).toMatchSnapshot();

		// loading
		await TestsEvent.triggerTouch(track, 'touchend', [[0, 100]]);

		expect(container).toMatchSnapshot();
	});

	it('should not emit onRefresh event after pulling a short distance', async () => {
		const onRefresh = jest.fn();

		const { container } = render(<PullRefresh onRefresh={onRefresh} />);

		const track = TestsDOM.mustQuerySelector(
			container,
			'.fnx-pull-refresh__track',
		);

		await TestsEvent.triggerDrag(track, [0, 10]);

		expect(onRefresh).not.toBeCalled();
	});

	it('should not trigger pull refresh when not in page top', async () => {
		const onRefresh = jest.fn();

		const { container } = render(<PullRefresh onRefresh={onRefresh} />);

		const track = TestsDOM.mustQuerySelector(
			container,
			'.fnx-pull-refresh__track',
		);

		// ignore touch event when not at page top
		await waitFor(() => {
			fireEvent.scroll(window, { target: { pageYOffset: 1 } });
		});

		await TestsEvent.triggerDrag(track, [0, 100]);
		expect(onRefresh).not.toBeCalled();

		await waitFor(() => {
			fireEvent.scroll(window, { target: { pageYOffset: 0 } });
		});

		await TestsEvent.triggerDrag(track, [0, 100]);
		expect(onRefresh).toBeCalled();
	});

	it('should render success text correctly', async () => {
		const onRefresh = jest.fn();

		const { container, rerender } = render(
			<PullRefresh
				slots={{ success: 'success' }}
				refreshing={true}
				onRefresh={onRefresh}
			/>,
		);

		const track = TestsDOM.mustQuerySelector(
			container,
			'.fnx-pull-refresh__track',
		);

		await TestsEvent.triggerDrag(track, [0, 100]);

		// loading
		expect(onRefresh).toBeCalled();

		await waitFor(() =>
			rerender(
				<PullRefresh
					slots={{ success: 'success' }}
					refreshing={false}
					onRefresh={onRefresh}
				/>,
			),
		);

		// success
		expect(container).toMatchSnapshot();

		act(() => {
			jest.runAllTimers();
		});

		// normal
		expect(container).toMatchSnapshot();
	});

	it('should set height when using indicatorHeight', async () => {
		const { container } = render(<PullRefresh indicatorHeight={100} />);

		const head = TestsDOM.mustQuerySelector<HTMLElement>(
			container,
			'.fnx-pull-refresh__indicator',
		);

		expect(head.style.height).toEqual('100px');
	});
});
