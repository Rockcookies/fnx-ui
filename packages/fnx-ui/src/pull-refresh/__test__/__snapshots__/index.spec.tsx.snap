// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PullRefresh/> should render different head content in different pulling status 1`] = `
<div>
  <div
    class="fnx-pull-refresh"
  >
    <div
      class="fnx-pull-refresh__track"
      style="transition-duration: 0ms; transform: translate3d(0, 20px, 0);"
    >
      <div
        class="fnx-pull-refresh__indicator"
      >
        <div
          class="fnx-pull-refresh__text"
        >
          Pull to refresh...
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<PullRefresh/> should render different head content in different pulling status 2`] = `
<div>
  <div
    class="fnx-pull-refresh"
  >
    <div
      class="fnx-pull-refresh__track"
      style="transition-duration: 0ms; transform: translate3d(0, 75px, 0);"
    >
      <div
        class="fnx-pull-refresh__indicator"
      >
        <div
          class="fnx-pull-refresh__text"
        >
          Loose to refresh...
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<PullRefresh/> should render different head content in different pulling status 3`] = `
<div>
  <div
    class="fnx-pull-refresh"
  >
    <div
      class="fnx-pull-refresh__track"
      style="transition-duration: 300ms; transform: translate3d(0, 50px, 0);"
    >
      <div
        class="fnx-pull-refresh__indicator"
      >
        <span
          aria-busy="true"
          aria-live="polite"
          class="fnx-loading"
        >
          <span
            class="fnx-icon"
            style="font-size: 16px;"
          >
            <svg
              aria-hidden="true"
              class="fnx-icon__spinner"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="25 25 50 50"
              width="1em"
            >
              <circle
                cx="50"
                cy="50"
                fill="none"
                r="20"
              />
            </svg>
          </span>
          <span
            class="fnx-loading__text"
          >
            Loading...
          </span>
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`<PullRefresh/> should render different head content in different pulling status 4`] = `
<div>
  <div
    class="fnx-pull-refresh"
  >
    <div
      class="fnx-pull-refresh__track"
      style="transition-duration: 300ms; transform: translate3d(0, 50px, 0);"
    >
      <div
        class="fnx-pull-refresh__indicator"
      >
        <span
          aria-busy="true"
          aria-live="polite"
          class="fnx-loading"
        >
          <span
            class="fnx-icon"
            style="font-size: 16px;"
          >
            <svg
              aria-hidden="true"
              class="fnx-icon__spinner"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="25 25 50 50"
              width="1em"
            >
              <circle
                cx="50"
                cy="50"
                fill="none"
                r="20"
              />
            </svg>
          </span>
          <span
            class="fnx-loading__text"
          >
            Loading...
          </span>
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`<PullRefresh/> should render different head content in different pulling status 5`] = `
<div>
  <div
    class="fnx-pull-refresh"
  >
    <div
      class="fnx-pull-refresh__track"
      style="transition-duration: 300ms;"
    >
      <div
        class="fnx-pull-refresh__indicator"
      />
    </div>
  </div>
</div>
`;

exports[`<PullRefresh/> should render status slots correctly 1`] = `
<div>
  <div
    class="fnx-pull-refresh"
  >
    <div
      class="fnx-pull-refresh__track"
      style="transition-duration: 0ms; transform: translate3d(0, 20px, 0);"
    >
      <div
        class="fnx-pull-refresh__indicator"
      >
        pulling 20
      </div>
    </div>
  </div>
</div>
`;

exports[`<PullRefresh/> should render status slots correctly 2`] = `
<div>
  <div
    class="fnx-pull-refresh"
  >
    <div
      class="fnx-pull-refresh__track"
      style="transition-duration: 0ms; transform: translate3d(0, 75px, 0);"
    >
      <div
        class="fnx-pull-refresh__indicator"
      >
        loosing 75
      </div>
    </div>
  </div>
</div>
`;

exports[`<PullRefresh/> should render status slots correctly 3`] = `
<div>
  <div
    class="fnx-pull-refresh"
  >
    <div
      class="fnx-pull-refresh__track"
      style="transition-duration: 300ms; transform: translate3d(0, 50px, 0);"
    >
      <div
        class="fnx-pull-refresh__indicator"
      >
        loading 50
      </div>
    </div>
  </div>
</div>
`;

exports[`<PullRefresh/> should render success text correctly 1`] = `
<div>
  <div
    class="fnx-pull-refresh"
  >
    <div
      class="fnx-pull-refresh__track"
      style="transition-duration: 300ms; transform: translate3d(0, 50px, 0);"
    >
      <div
        class="fnx-pull-refresh__indicator"
      >
        success
      </div>
    </div>
  </div>
</div>
`;

exports[`<PullRefresh/> should render success text correctly 2`] = `
<div>
  <div
    class="fnx-pull-refresh"
  >
    <div
      class="fnx-pull-refresh__track"
      style="transition-duration: 300ms;"
    >
      <div
        class="fnx-pull-refresh__indicator"
      />
    </div>
  </div>
</div>
`;
